hedera-node/docs/proposals/block-stream-bucket-uploader.md

```diff
@@ -0,0 +1,398 @@
+    @ConfigProperty String region,    // required for AWS only
+    @ConfigProperty String bucketName
+) {
+    public BucketNetworkConfig {
+        Objects.requireNonNull(name, "Bucket name cannot be null");
+        Objects.requireNonNull(provider, "Provider cannot be null");
+        Objects.requireNonNull(endpoint, "Endpoint cannot be null");
+        Objects.requireNonNull(bucketName, "Bucket name cannot be null");
+        
+        if ("aws".equals(provider)) {
+            Objects.requireNonNull(region, "Region is required for AWS buckets");
+        }
+    }
+}
+```
+
+#### Credentials Configuration
+The credentials are stored in a separate JSON file on disk for security:
+
+```json
+{
+  "credentials": {
+    "default-aws-bucket": {
+      "accessKey": "YOUR_AWS_ACCESS_KEY",
+      "secretKey": "YOUR_AWS_SECRET_KEY"
+    },
+    "default-gcp-bucket": {
+      "accessKey": "YOUR_GCP_ACCESS_KEY",
+      "secretKey": "YOUR_GCP_SECRET_KEY"
+    }
+  }
+}
+```
+
+#### Credentials Records
+```java
+public record BucketCredentials(
+    String accessKey,
+    String secretKey
+) {
+    public BucketCredentials {
+        Objects.requireNonNull(accessKey, "Access key cannot be null");
+        Objects.requireNonNull(secretKey, "Secret key cannot be null");
+    }
+}
+
+public record BucketCredentialsConfig(
+    Map<String, BucketCredentials> credentials
+) {
+    public BucketCredentialsConfig {
+        Objects.requireNonNull(credentials, "Credentials map cannot be null");
+    }
+}
+```
+
+#### Configuration Loading and Updates
+The configuration system needs to handle both network properties and local credentials in a secure and maintainable way.
+
+**BucketConfigurationManager Requirements**
+- Singleton class managing both network properties and local credentials
+- Load credentials from the JSON file specified in the configuration
+- Provide methods to get complete bucket configurations by combining network properties with credentials
+- Handle credential file reload if needed
+- Log appropriate warnings/errors when credentials are missing or invalid
+- Validate that all required bucket configurations have matching credentials
+
+**CompleteBucketConfig Requirements**
+- Immutable record combining network configuration with credentials
+- Include all necessary fields for bucket connection:
+  - name
+  - provider type (AWS/GCP)
+  - endpoint
+  - region (if applicable)
+  - bucket name
+  - credentials (access key and secret key)
+- Include validation to ensure all required fields are present
+- Provide a clean interface for the MinioClientFactory to create new clients
+
+### Cloud Storage Interface and Implementations
+
+#### CloudBucketUploader Interface
+The base interface defining operations for uploading blocks to cloud storage providers.
+
+Key Operations:
+- Upload a block file with associated metadata asynchronously
+- Check if a block exists in cloud storage
+- Retrieve MD5 hash of a stored block
+- Get the provider identifier
```

Comment:
What is the provider identifier and what does it tell us? I think we should have separate methods for calling AWS and GCS so we should know the provider up front.